### PR TITLE
[FLINK-12784][metrics] Support retention policy for InfluxDB metrics …

### PR DESCRIPTION
**Pull request by mans2singh**
## What is the purpose of the change

* Support optional retention policy for InfluxDB metric reporter (currently only default is allowed)

## Brief change log

* Added support for retention policy specification for InfluxDB metrics reporter along with supporting tests and documentation.

## Verifying this change

This change added tests and can be verified as follows:

* Running the unit test suite

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: docs/monitoring/metrics.md [hunk 1]
```diff
@@ -658,6 +658,7 @@ Parameters:
 - `db` - the InfluxDB database to store metrics
 - `username` - (optional) InfluxDB username used for authentication
 - `password` - (optional) InfluxDB username's password used for authentication
+- `rp` - (optional) InfluxDB retention policy, defaults to retention policy defined on the server
```
Comment (Myasuka) [inline docs/monitoring/metrics.md]: Updates in `metrics.md` should also be updated in `metrics-zh.md`. Moreover, `metrics.hmtl` should also be updated, please refer to `flink-docs/README.md` to generate docs.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java [hunk 2]
```diff
@@ -57,6 +57,7 @@
 	private static final String TEST_INFLUXDB_DB = "test-42";
 	private static final String METRIC_HOSTNAME = "task-mgr-1";
 	private static final String METRIC_TM_ID = "tm-id-123";
+	private String retentionPolicy = "";
```
Comment (Myasuka) [inline flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: How about using final variable `RETENTION_POLICY` just like above tests?

File: docs/monitoring/metrics.md (hunk 1 above)
Comment (mans2singh) [reply docs/monitoring/metrics.md]: You are right @Myasuka - the retention policy needs to created when creating the database.  Also, let me now what's the process the other docs you've mentioned.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java (hunk 2 above)
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: I did use final variable for retention policy in the test case so that the createMetricsRegistry can return the registry impl with  properties both default and non-default test  scenrios.   Please let me now if you have any additiona advice.  Thanks

File: docs/monitoring/metrics.md (hunk 1 above)
Comment (Myasuka) [reply docs/monitoring/metrics.md]: Please refer to [generate config doc](https://github.com/apache/flink/blob/master/flink-docs/README.md#configuration-documentation) to know the process. And you could search the `metrics-zh.md` to make they are the same as English version.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java (hunk 2 above)
Comment (1u0) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: Just personal opinion: I think the second test (`testMetricReportingWithRetentionPolicy`, as it is now) is redundant, and doesn't show some new aspect under test. Alternatives could be, to keep one test and either:
 * make retention policy parameter as `private static final` (with some random, non-empty string);
 * OR, make the `retentionPolicy` variable local in the test method and explicitly pass it to `createMetricRegistry()`.

File: docs/monitoring/metrics.md (hunk 1 above)
Comment (mans2singh) [reply docs/monitoring/metrics.md]: I've updated the metrics.zh.md - thanks for the pointer.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java (hunk 2 above)
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: I've added the retention policy argument to createMetricRegistry method and updated tests.

File: flink-metrics/flink-metrics-influxdb/src/main/java/org/apache/flink/metrics/influxdb/InfluxdbReporter.java [hunk 3]
```diff
@@ -53,6 +54,7 @@
 
 	private String database;
 	private InfluxDB influxDB;
+	private String retentionPolicy;
```
[the remark this answers is not in the document]
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/main/java/org/apache/flink/metrics/influxdb/InfluxdbReporter.java]: @1u0 Updated based on your recommendation.

File: flink-metrics/flink-metrics-influxdb/src/main/java/org/apache/flink/metrics/influxdb/InfluxdbReporterOptions.java [hunk 4]
```diff
@@ -52,6 +52,11 @@
 		.noDefaultValue()
 		.withDescription("the InfluxDB database to store metrics");
 
+	public static final ConfigOption<String> RETENTION_POLICY = ConfigOptions
+			.key("rp")
```
[the remark this answers is not in the document]
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/main/java/org/apache/flink/metrics/influxdb/InfluxdbReporterOptions.java]: @1u0 - Updated to full name based on your recommendation.

File: flink-metrics/flink-metrics-influxdb/src/main/java/org/apache/flink/metrics/influxdb/InfluxdbReporter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (1u0) [reply flink-metrics/flink-metrics-influxdb/src/main/java/org/apache/flink/metrics/influxdb/InfluxdbReporter.java]: What I meant is to reorder the fields, like:
```java
private String database;
private String retentionPolicy;
private InfluxDB influxDB;
```
and the same can be done for initialization order in `open()` method.

[the remark this answers is not in the document]
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/main/java/org/apache/flink/metrics/influxdb/InfluxdbReporter.java]: @1u0  - I've updated the order of fields and initialization.  Please let me know your thoughts.

**Comment by mans2singh** [comment]
@1u0 - I've updated the test case (consolidated the testMetricReporting into one).  Please let me know if you have any additional advice.  Thanks again for your feedback.

**Comment by mans2singh** [comment]
@1u0 @Myasuka @rmetzger The python build is failing with the error 

> The command "./tools/travis_controller.sh python" exited with 1.

for this PR [build failure](https://travis-ci.org/apache/flink/jobs/544979681).  Can you please let me know how to resolve this issue ?  

Also, let me know if you have any other recommendations.

**Comment by Myasuka** [comment]
The root cause is `Cached flink dir /home/travis/flink_cache/37549/flink does not exist. Exiting build`. From my experience, this might be a travis bug, as far as I know, you could try to close - reopen this PR or create a new commit to retrigger the build phase.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java [hunk 5]
```diff
@@ -96,33 +96,36 @@ public void testMetricRegistration() throws Exception {
...
 			reporter.report();
 
 			verify(postRequestedFor(urlPathEqualTo("/write"))
-				.withQueryParam("db", equalTo(TEST_INFLUXDB_DB))
-				.withHeader("Content-Type", containing("text/plain"))
-				.withRequestBody(containing("taskmanager_" + metricName + ",host=" + METRIC_HOSTNAME + ",tm_id=" + METRIC_TM_ID + " count=42i")));
+					.withQueryParam("db", equalTo(TEST_INFLUXDB_DB))
+					.withQueryParam("rp", equalTo(retentionPolicy))
+					.withHeader("Content-Type", containing("text/plain"))
+					.withRequestBody(containing("taskmanager_" + metricName + ",host=" + METRIC_HOSTNAME + ",tm_id=" + METRIC_TM_ID + " count=42i")));
```
Comment (1u0) [inline flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: The code formatting has changed here and 10 lines above.

Otherwise, I don't have anything to add any more.
@mans2singh, you can resolve my other comments if you want.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java [hunk 6]
```diff
@@ -96,33 +96,36 @@ public void testMetricRegistration() throws Exception {
...
 		try {
 			String metricName = "TestCounter";
 			Counter counter = registerTestMetric(metricName, metricRegistry);
 			counter.inc(42);
 
 			stubFor(post(urlPathEqualTo("/write"))
-				.willReturn(aResponse()
-					.withStatus(200)));
+					.willReturn(aResponse()
+							.withStatus(200)));
```
Comment (1u0) [inline flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: > The code formatting has changed here and 10 lines above.

Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: @1u0 - Updated formatting based on your recommendation.  Please let me know if have any more comments.  Thanks again.

Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: @1u0, @Myasuka - Please let me know if I can resolve this issue and if there is anything else required for this PR.  Thanks for your advice.

Mans

File: docs/monitoring/metrics.md [hunk 7]
```diff
@@ -669,10 +670,11 @@ metrics.reporter.influxdb.port: 8086
 metrics.reporter.influxdb.db: flink
 metrics.reporter.influxdb.username: flink-metrics
 metrics.reporter.influxdb.password: qwerty
+metrics.reporter.influxdb.retentionPolicy: one_hr
 
 {% endhighlight %}
 
-The reporter would send metrics using http protocol with default retention policy defined on InfluxDB server.
+The reporter would send metrics using http protocol to the InfluxDB server with the specified retention policy (or the default policy specified on the server).
```
[the remark this answers is not in the document]
Comment (mans2singh) [reply docs/monitoring/metrics.md]: @1u0 - Corrected documentation.  Thanks again for your advice.

**Comment by mans2singh** [comment]
@zentol @rmetzger - Please let me know if you can review this request or point me to someone who can advice on it.  BTW - The travis build failure is not related to this enhancement and it's test cases.  Thanks

**Comment by mans2singh** [comment]
@1u0 @Myasuka - Can you please advice how I can get this PR reviewed by someone who has merge privs ?  Thanks

**Comment by Myasuka** [comment]
@mans2singh Travis seems not so stable, see [mail](http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/Something-wrong-with-travis-td29683.html). I'm afraid you have to retrigger this travis again. Since committers might have limited resource to review PRs, if ping him via github works not so well, you could try to send email to them directly.

**Comment by mans2singh** [comment]
@Myasuka - Thanks for your approval.  I am waiting for travis to stablize and will then push a commit.  Thanks again for your help.

File: docs/monitoring/metrics.md [hunk 8]
```diff
@@ -669,10 +670,11 @@ metrics.reporter.influxdb.port: 8086
 metrics.reporter.influxdb.db: flink
 metrics.reporter.influxdb.username: flink-metrics
 metrics.reporter.influxdb.password: qwerty
+metrics.reporter.influxdb.retentionPolicy: one_hr
```
Comment (NicoK) [inline docs/monitoring/metrics.md]: I know, this is a user-defined policy name, but how about using "one_hour" as in the tests?

File: docs/monitoring/metrics.zh.md [hunk 9]
```diff
@@ -667,10 +668,11 @@ metrics.reporter.influxdb.port: 8086
 metrics.reporter.influxdb.db: flink
 metrics.reporter.influxdb.username: flink-metrics
 metrics.reporter.influxdb.password: qwerty
+metrics.reporter.influxdb.retentionPolicy: one_hr
```
Comment (NicoK) [inline docs/monitoring/metrics.zh.md]: same as above: use "one_hour" as in the tests?

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java [hunk 10]
```diff
@@ -66,7 +66,7 @@
 
 	@Test
 	public void testReporterRegistration() throws Exception {
-		MetricRegistryImpl metricRegistry = createMetricRegistry();
+		MetricRegistryImpl metricRegistry = createMetricRegistry("");
```
Comment (NicoK) [inline flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: use `RETENTION_POLICY.defaultValue()`?

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java [hunk 11]
```diff
@@ -78,7 +78,7 @@ public void testReporterRegistration() throws Exception {
 
 	@Test
 	public void testMetricRegistration() throws Exception {
-		MetricRegistryImpl metricRegistry = createMetricRegistry();
+		MetricRegistryImpl metricRegistry = createMetricRegistry("");
```
Comment (NicoK) [inline flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: use `RETENTION_POLICY.defaultValue()`?

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java [hunk 12]
```diff
@@ -111,18 +112,20 @@ public void testMetricReporting() throws Exception {
...
 		}
 	}
 
-	private MetricRegistryImpl createMetricRegistry() {
+	private MetricRegistryImpl createMetricRegistry(String retentionPolicy) {
 		MetricConfig metricConfig = new MetricConfig();
 		metricConfig.setProperty("host", "localhost");
 		metricConfig.setProperty("port", String.valueOf(wireMockRule.port()));
 		metricConfig.setProperty("db", TEST_INFLUXDB_DB);
+		metricConfig.setProperty("retentionPolicy", retentionPolicy);
```
Comment (NicoK) [inline flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: Actually, it would be nicer to have all these config options use the keys from the `InfluxdbReporterOptions`, e.g. `HOST.key()` or `RETENTION_POLICY.key()` so that they are up-to date if options are renamed/deprecated.

File: docs/monitoring/metrics.md (hunk 8 above)
Comment (mans2singh) [reply docs/monitoring/metrics.md]: Updated.

File: docs/monitoring/metrics.zh.md (hunk 9 above)
Comment (mans2singh) [reply docs/monitoring/metrics.zh.md]: Updated based on recommendation.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java (hunk 10 above)
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: Corrected.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java (hunk 11 above)
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: Corrected.

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java (hunk 12 above)
Comment (mans2singh) [reply flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: Corrected based on recommendations.

**Comment by mans2singh** [comment]
@NicoK - 

Thanks for your recommendations. 

I've updated the code and documentation based on your feedback.  One note that for the retention policy, the request parameter key in [test case](https://github.com/apache/flink/pull/8668/files#diff-4ab6d3fdde78854f34e1006ee8335812R115) is 'rp' as required by InfluxDB even though the property name is 'retentionPolicy' for readability. 

Please let me know if you have other recommendations and if I can resolve the comments.

Thanks again for your time.

Mans

**Comment by mans2singh** [comment]
@NicoK 

The travis build is failing for connectors and python.

Here is the error from [connectors failure](https://travis-ci.org/apache/flink/jobs/551049381)

> 05:57:01.838 [ERROR] Tests run: 10, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.147 s <<< FAILURE! - in org.apache.flink.streaming.connectors.kinesis.FlinkKinesisProducerTest
05:57:01.838 [ERROR] testBackpressure(org.apache.flink.streaming.connectors.kinesis.FlinkKinesisProducerTest)  Time elapsed: 0.543 s  <<< FAILURE!
java.lang.AssertionError: Flush triggered before reaching queue limit
	at org.apache.flink.streaming.connectors.kinesis.FlinkKinesisProducerTest.testBackpressure(FlinkKinesisProducerTest.java:298)

Also [travis python build](https://travis-ci.org/apache/flink/jobs/551049378) is failing which are apparently not related to this pull request

Please let me know if you have any advice for me.

Thanks.

Mans

File: flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java [hunk 13]
```diff
@@ -111,7 +111,7 @@ public void testMetricReporting() throws Exception {
 			reporter.report();
 
 			verify(postRequestedFor(urlPathEqualTo("/write"))
-				.withQueryParam("db", equalTo(TEST_INFLUXDB_DB))
+				.withQueryParam(InfluxdbReporterOptions.DB.key(), equalTo(TEST_INFLUXDB_DB))
```
Comment (NicoK) [inline flink-metrics/flink-metrics-influxdb/src/test/java/org/apache/flink/metrics/influxdb/InfluxdbReporterTest.java]: Please revert this change here, since the query param is actually unrelated to how we name our options. In my last review, I was only referring to the `createMetricRegistry` method below.
